### ISSUE TITLE
Make login workflow consistent with user expecations

### ISSUE THREAD

**Issue by FaeyUmbrea** [foundryvtt/foundryvtt#8512]
### User Experience

When opening foundry URLs that require login, there is a redirect to a login page if the requred URL is unauthorized.

This works fine for 95% of situations within foundry, however there is inconsistency with the /stream page.

When opening /stream you are redirected to /join, and then, upon login, redirected to /game. Which breaks the usual redirect - login - return flow that users have become familiar with.

It would be great if the redirect attached the return URL in the generated request and the Login Form would send the user back to the URL they came from.

**Comment by aaclayton** [comment]
Fair enough, although the `/stream` page is quite an exception or special case. There is only one view which a normal user goes to after logging in to the VTT and that is `/game`. Tracking where the user was previously redirected from so that they can be sent to their desired route after authentication is - in my opinion - only of _minimal_ use.

**Comment by FaeyUmbrea** [comment]
I do agree here, that there probably isn't a lot of people who end up utilizing this route, but it would definitely help the people using Foundry for Streaming.

**Comment by aaclayton** [comment]
Comment from @numinit 

### User Experience

When /game redirects to /join for login, URL query arguments should be preserved.

Expected: visiting `http://my.foundry:30000/game?param=value` without login redirects to `http://my.foundry:30000/join?param=value`

Actual behavior: visiting `http://my.foundry:30000/game?param=value` without login redirects to `http://my.foundry:30000/join`

This enables deep linking usecases such as https://github.com/numinit/foundry-permalinks without external help from nginx.

(Note that there is a related bug - the `<body>` element should have the class name `game` when there are query params present, but actually has the class name `game?param=value`).

**Comment by numinit** [comment]
For what it's worth, if you can run module code on the page context, you can grab the document.referrer instead of relying on the query params being preserved through to /game:

https://github.com/numinit/foundry-permalinks/blob/master/src/index.ts#L109

This probably reduces the lift on Foundry's codebase quite a bit anyway, since all that's needed is preserving query params on the 302 redirect to /join for this particular usecase - nothing else.

**Comment by FaeyUmbrea** [comment]
That only helps with anything that runs inside /game but since /join will never redirect to /stream it won't solve the initial problem discussed in this issue.

**Comment by numinit** [comment]
Ah, OK, so it would require another query param like you suggested. I see.
